cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,574 @@
+    handled by the autoscalers. In case the buffer chunks were very small the
+    number of processed pods may grow - we do not expect that buffers would
+    introduce more chunks than the number of pods that user may want to run.
+    Since these are only virtual they will introduce less overhead in the
+    cluster than balloon pods/deployments (no need to preempt etc).
+
+# Design Details
+
+## Proposed API
+
+```go
+type NodeAutoscalingBuffer struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Spec   NodeAutoscalingBufferSpec
+    Status NodeAutoscalingBufferStatus
+}
+
+type NodeAutoscalingBufferSpec struct {
+    Capacity  NodeAutoscalingBufferCapacity
+
+    // Initially only active stand by meaning scaling up the cluster will be available.
+    // Not specified strategy will mean using buffer.x-k8s.io/active-capacity	
+    // In the future the cloud providers will be able to offer their own buffering strategies.
+    ProvisioningStrategy string
+}
+
+type NodeAutoscalingBufferCapacity struct {
+    // One of the following must be set
+    // +optional
+    Replicas *ReplicaBufferCapacity
+    // +optional
+    Resources *ResourceBufferCapacity
+    // +optional
+    Pods *PodBufferCapacity
+}
+
+type ReplicaBufferCapacity struct {
+    // An object implementing scale subresource and setting its label selector
+    // field (for example deployment, job etc.)
+    TargetRef *TypedLocalObjectReference
+    // One of the following must be set:
+    // +optional
+    Exactly *int
+    // +optional
+    Percent *PercentReplicaBufferCapacity
+}
+
+type PercentReplicaBufferCapacity struct {
+    Percent  int
+    MinCount *int
+    MaxCount *int
+}
+
+type ResourceBufferCapacity struct {
+    // Defines where the resource should be created.
+    TargetRef *TypedLocalObjectReference
+    // Upper limit of the total capacity that will be scaled for this buffer.
+    // The total capacity will be split into smaller chunks that each fits onto nodes.
+    // The size of a chunk will be either decided for the user (using some defaults and heuristics) or the
+    // user can specify per pod Requests to define the shape of the buffer space. In the latter case the
+    // CPU and Mem will be the maximum size of the buffer and if the Requests are too big to fit the total
+    // buffer size the admission of the buffer should be denied.
+    TotalCpu resource.Quantity
+    TotalMem resource.Quantity
+
+    // Optional if defining a single chunk size is needed. The controller will translate this into enough
+    // pods to create buffer not exceeding the total size above
+    PerChunkResources *ResourceLimits
+    // Optional if more detailed limitations than target object ref on its own is not enough.
+    NodeSelector NodeSelector
+}
+
+type PodBufferCapacity struct {
+    ReplicaNum int
+    PodTemplate PodTemplateRef
+}
+
+type NodeAutoscalingBufferStatus struct {
+    PodBufferCapacity   PodBufferCapacity
+	PodTemplateGeneration int	
```

Comment:
nit: some indents on the API are not aligned well, I think because of mixing tabs with spaces.